core/src/main/java/org/elasticsearch/monitor/jvm/JvmStats.java

```diff
@@ -94,55 +88,84 @@ public static JvmStats jvmStats() {
-            stats.bufferPools = new ArrayList<>(bufferPools.size());
+            bufferPoolsList = new ArrayList<>(bufferPools.size());
             for (BufferPoolMXBean bufferPool : bufferPools) {
-                stats.bufferPools.add(new BufferPool(bufferPool.getName(), bufferPool.getCount(), bufferPool.getTotalCapacity(), bufferPool.getMemoryUsed()));
+                bufferPoolsList.add(new BufferPool(bufferPool.getName(), bufferPool.getCount(),
+                        bufferPool.getTotalCapacity(), bufferPool.getMemoryUsed()));
             }
         } catch (Exception e) {
             // buffer pools are not available
         }
 
-        stats.classes = new Classes();
-        stats.classes.loadedClassCount = classLoadingMXBean.getLoadedClassCount();
-        stats.classes.totalLoadedClassCount = classLoadingMXBean.getTotalLoadedClassCount();
-        stats.classes.unloadedClassCount = classLoadingMXBean.getUnloadedClassCount();
+        Classes classes = new Classes(classLoadingMXBean.getLoadedClassCount(), classLoadingMXBean.getTotalLoadedClassCount(),
+                classLoadingMXBean.getUnloadedClassCount());
 
-        return stats;
+        return new JvmStats(System.currentTimeMillis(), runtimeMXBean.getUptime(), mem, threads,
+                garbageCollectors, bufferPoolsList, classes);
     }
 
-    long timestamp = -1;
-    long uptime;
-    Mem mem;
-    Threads threads;
-    GarbageCollectors gc;
-    List<BufferPool> bufferPools;
-    Classes classes;
-
-    private JvmStats() {
-    }
+    private final long timestamp;
+    private final long uptime;
+    private final Mem mem;
+    private final Threads threads;
+    private final GarbageCollectors gc;
+    private final List<BufferPool> bufferPools;
+    private final Classes classes;
 
-    public JvmStats(long timestamp, long uptime) {
+    public JvmStats(long timestamp, long uptime, Mem mem, Threads threads, GarbageCollectors gc,
+                    @Nullable List<BufferPool> bufferPools, Classes classes) {
         this.timestamp = timestamp;
         this.uptime = uptime;
+        this.mem = mem;
+        this.threads = threads;
+        this.gc = gc;
+        this.bufferPools = bufferPools;
+        this.classes = classes;
+    }
+
+    public JvmStats(StreamInput in) throws IOException {
+        timestamp = in.readVLong();
+        uptime = in.readVLong();
+        mem = new Mem(in);
+        threads = new Threads(in);
+        gc = new GarbageCollectors(in);
+        if (in.readBoolean()) {
+            bufferPools = in.readList(BufferPool::new);
+        } else {
+            bufferPools = null;
+        }
+        classes = new Classes(in);
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeVLong(timestamp);
+        out.writeVLong(uptime);
+        mem.writeTo(out);
+        threads.writeTo(out);
+        gc.writeTo(out);
+        if (bufferPools == null) {
```

Comment:
Maybe we should just use `emptyList()` instead? That way we don't need the conditional.